File: windows/GreatPush.Shared/SettingsManager.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
#if WINDOWS_UWP
using Windows.Storage;
#else
using GreatPush.Net46.Adapters.Storage;
using System.IO;
#endif

namespace GreatPush.ReactNative
{
    internal class SettingsManager
    {
        private static ApplicationDataContainer Settings = null;

        static SettingsManager ()
        {

#if WINDOWS_UWP
            Settings = ApplicationData.Current.LocalSettings.CreateContainer(GreatPushConstants.GreatPushPreferences, ApplicationDataCreateDisposition.Always);
#else
            var folder = UpdateUtils.GetGreatPushFolderAsync().Result;
            Settings = new ApplicationDataContainer(Path.Combine(folder.Path, GreatPushConstants.GreatPushPreferences));
#endif
        }

        public static JArray GetFailedUpdates()
        {
            var failedUpdatesString = (string)Settings.Values[GreatPushConstants.FailedUpdatesKey];
            if (failedUpdatesString == null)
            {
                return new JArray();
            }

            try
            {
                return JArray.Parse(failedUpdatesString);
            }
            catch (Exception)
            {
                var emptyArray = new JArray();
                Settings.Values[GreatPushConstants.FailedUpdatesKey] = JsonConvert.SerializeObject(emptyArray);
                return emptyArray;
            }
        }

        internal static JObject GetPendingUpdate()
        {
            var pendingUpdateString = (string)Settings.Values[GreatPushConstants.PendingUpdateKey];
            if (pendingUpdateString == null)
            {
                return null;
            }

            try
            {
                return JObject.Parse(pendingUpdateString);
            }
            catch (Exception)
            {
                // Should not happen.
                GreatPushUtils.Log("Unable to parse pending update metadata " + pendingUpdateString +
                        " stored in SharedPreferences");
                return null;
            }
        }

        internal static bool IsFailedHash(string packageHash)
        {
            JArray failedUpdates = SettingsManager.GetFailedUpdates();
            if (packageHash != null)
            {
                foreach (var failedPackage in failedUpdates)
                {
                    var failedPackageHash = (string)failedPackage[GreatPushConstants.PackageHashKey];
                    if (packageHash.Equals(failedPackageHash))
                    {
                        return true;
                    }
                }
            }

            return false;
        }

        internal static bool IsPendingUpdate(string packageHash)
        {
            JObject pendingUpdate = SettingsManager.GetPendingUpdate();
            return pendingUpdate != null &&
                    !(bool)pendingUpdate[GreatPushConstants.PendingUpdateIsLoadingKey] &&
                    (packageHash == null || ((string)pendingUpdate[GreatPushConstants.PendingUpdateHashKey]).Equals(packageHash));
        }

        internal static void RemoveFailedUpdates()
        {
            Settings.Values.Remove(GreatPushConstants.FailedUpdatesKey);
        }

        internal static void RemovePendingUpdate()
        {
            Settings.Values.Remove(GreatPushConstants.PendingUpdateKey);
        }

        internal static void SaveFailedUpdate(JObject failedPackage)
        {
            var failedUpdatesString = (string)Settings.Values[GreatPushConstants.FailedUpdatesKey];
            JArray failedUpdates;
            if (failedUpdatesString == null)
            {
                failedUpdates = new JArray();
            }
            else
            {
                failedUpdates = JArray.Parse(failedUpdatesString);
            }

            failedUpdates.Add(failedPackage);
            Settings.Values[GreatPushConstants.FailedUpdatesKey] = JsonConvert.SerializeObject(failedUpdates);
        }

        internal static void SavePendingUpdate(string packageHash, bool isLoading)
        {
            var pendingUpdate = new JObject()
            {
                { GreatPushConstants.PendingUpdateHashKey, packageHash },
                { GreatPushConstants.PendingUpdateIsLoadingKey, isLoading }
            };

            Settings.Values[GreatPushConstants.PendingUpdateKey] = JsonConvert.SerializeObject(pendingUpdate);
        }

        internal static void SetString(string key, string value)
        {
            Settings.Values[key] = value;
        }

        internal static string GetString(string key)
        {
            return (string)Settings.Values[key];
        }

        internal static void RemoveString(string key)
        {
            Settings.Values.Remove(key);
        }
    }
}

File: android/app/src/main/java/com/microsoft/greatpush/react/SettingsManager.java
package com.microsoft.greatpush.react;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class SettingsManager {

    private SharedPreferences mSettings;

    public SettingsManager(Context applicationContext) {
        mSettings = applicationContext.getSharedPreferences(GreatPushConstants.GREAT_PUSH_PREFERENCES, 0);
    }

    public JSONArray getFailedUpdates() {
        String failedUpdatesString = mSettings.getString(GreatPushConstants.FAILED_UPDATES_KEY, null);
        if (failedUpdatesString == null) {
            return new JSONArray();
        }

        try {
            return new JSONArray(failedUpdatesString);
        } catch (JSONException e) {
            // Unrecognized data format, clear and replace with expected format.
            JSONArray emptyArray = new JSONArray();
            mSettings.edit().putString(GreatPushConstants.FAILED_UPDATES_KEY, emptyArray.toString()).commit();
            return emptyArray;
        }
    }

    public JSONObject getPendingUpdate() {
        String pendingUpdateString = mSettings.getString(GreatPushConstants.PENDING_UPDATE_KEY, null);
        if (pendingUpdateString == null) {
            return null;
        }

        try {
            return new JSONObject(pendingUpdateString);
        } catch (JSONException e) {
            // Should not happen.
            GreatPushUtils.log("Unable to parse pending update metadata " + pendingUpdateString +
                    " stored in SharedPreferences");
            return null;
        }
    }


    public boolean isFailedHash(String packageHash) {
        JSONArray failedUpdates = getFailedUpdates();
        if (packageHash != null) {
            for (int i = 0; i < failedUpdates.length(); i++) {
                try {
                    JSONObject failedPackage = failedUpdates.getJSONObject(i);
                    String failedPackageHash = failedPackage.getString(GreatPushConstants.PACKAGE_HASH_KEY);
                    if (packageHash.equals(failedPackageHash)) {
                        return true;
                    }
                } catch (JSONException e) {
                    throw new GreatPushUnknownException("Unable to read failedUpdates data stored in SharedPreferences.", e);
                }
            }
        }

        return false;
    }

    public boolean isPendingUpdate(String packageHash) {
        JSONObject pendingUpdate = getPendingUpdate();

        try {
            return pendingUpdate != null &&
                    !pendingUpdate.getBoolean(GreatPushConstants.PENDING_UPDATE_IS_LOADING_KEY) &&
                    (packageHash == null || pendingUpdate.getString(GreatPushConstants.PENDING_UPDATE_HASH_KEY).equals(packageHash));
        }
        catch (JSONException e) {
            throw new GreatPushUnknownException("Unable to read pending update metadata in isPendingUpdate.", e);
        }
    }

    public void removeFailedUpdates() {
        mSettings.edit().remove(GreatPushConstants.FAILED_UPDATES_KEY).commit();
    }

    public void removePendingUpdate() {
        mSettings.edit().remove(GreatPushConstants.PENDING_UPDATE_KEY).commit();
    }

    public void saveFailedUpdate(JSONObject failedPackage) {
        String failedUpdatesString = mSettings.getString(GreatPushConstants.FAILED_UPDATES_KEY, null);
        JSONArray failedUpdates;
        if (failedUpdatesString == null) {
            failedUpdates = new JSONArray();
        } else {
            try {
                failedUpdates = new JSONArray(failedUpdatesString);
            } catch (JSONException e) {
                // Should not happen.
                throw new GreatPushMalformedDataException("Unable to parse failed updates information " +
                        failedUpdatesString + " stored in SharedPreferences", e);
            }
        }

        failedUpdates.put(failedPackage);
        mSettings.edit().putString(GreatPushConstants.FAILED_UPDATES_KEY, failedUpdates.toString()).commit();
    }

    public void savePendingUpdate(String packageHash, boolean isLoading) {
        JSONObject pendingUpdate = new JSONObject();
        try {
            pendingUpdate.put(GreatPushConstants.PENDING_UPDATE_HASH_KEY, packageHash);
            pendingUpdate.put(GreatPushConstants.PENDING_UPDATE_IS_LOADING_KEY, isLoading);
            mSettings.edit().putString(GreatPushConstants.PENDING_UPDATE_KEY, pendingUpdate.toString()).commit();
        } catch (JSONException e) {
            // Should not happen.
            throw new GreatPushUnknownException("Unable to save pending update.", e);
        }
    }

}
